department-of-veterans-affairs/abd-vro
# lib-hoppy: Update to include per message TTL parameter for AsyncHoppyClient.

## What was the problem?

During a recent BGS outage, 3 requests for two failed jobs (6 total) were sent it with no response. Currently, there is no queue wide message TTL policy, and no per message TTL was being set by the publisher (AsyncHoppyClient) when sending the requests, and although the service was down, the messages sent to add notes to a claim were still in queue. When the service came back up, it resulted in duplicate notes being added to the queue.

Associated tickets or Slack threads:
- #3175 

## How does this fix it?[^1]

Adds a configurable per-message TTL to the message upon [publishing](https://www.rabbitmq.com/docs/ttl#per-message-ttl-in-publishers). The default is set to zero. From the documentation:

>  Setting the TTL to 0 causes messages to be expired upon reaching a queue unless they can be delivered to a consumer immediately. Thus this provides an alternative to the immediate publishing flag, which the RabbitMQ server does not support. Unlike that flag, no basic.returns are issued, and if a dead letter exchange is set then messages will be dead-lettered.

There is currently no dead lettering happening, and is unsupported by VRO and 

## How to test this PR
- navigate to hoppy folder
  - `cd abd-vro/shared/lib-hoppy`
- install new hoppy lib (in virtual env if you have it set up)
  - `pip install .`
- run `pytest`  

[^1]: [Pull-Requests guidelines](https://github.com/department-of-veterans-affairs/abd-vro/wiki/Pull-Requests). If PR is significant, update [Current Software State](https://github.com/department-of-veterans-affairs/abd-vro/wiki/Current-Software-State) wiki page.
[^secrel]: To check if a PR will succeed in the SecRel workflow, [test PRs in the SecRel pipeline](https://github.com/department-of-veterans-affairs/abd-vro-internal/wiki/Secure-Release-process#to-test-prs-in-the-secrel-pipeline).
